kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -0,0 +1,997 @@
+                            NumberOfControlPointsU(), NumberOfControlPointsV(), NumberOfControlPointsW(),
+                                cp_index_u, cp_index_v, cp_index_w);
+
+                        if (u == 0 && v==0 && w==0)
+                            rGlobalSpaceDerivatives[shape_function_row_i] =
+                            (*this)[index] * shape_function_container(u, v, w, shape_function_row_i);
+                        else
+                            rGlobalSpaceDerivatives[shape_function_row_i] +=
+                            (*this)[index] * shape_function_container(u, v, w, shape_function_row_i);
+                    }
+                }
+            }
+        }
+    }
+
+    ///@}
+    ///@name Shape Function
+    ///@{
+
+    Vector& ShapeFunctionsValues(
+        Vector &rResult,
+        const CoordinatesArrayType& rCoordinates) const override
+    {
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 0);
+
+        if (IsRational()) {
+            shape_function_container.ComputeNurbsShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW, mWeights,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+        else {
+            shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+
+        if (rResult.size() != shape_function_container.NumberOfNonzeroControlPoints())
+            rResult.resize(shape_function_container.NumberOfNonzeroControlPoints());
+
+        for (IndexType i = 0; i < shape_function_container.NumberOfNonzeroControlPoints(); ++i) {
+            rResult[i] = shape_function_container(i, 0);
+        }
+
+        return rResult;
+    }
+
+    Matrix& ShapeFunctionsLocalGradients(
+        Matrix& rResult,
+        const CoordinatesArrayType& rCoordinates) const override
+    {
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 1);
+
+        if (IsRational()) {
+            shape_function_container.ComputeNurbsShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW, mWeights,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+        else {
+            shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+
+        if (rResult.size1() != shape_function_container.NumberOfNonzeroControlPoints()
+            && rResult.size2() != 3)
+            rResult.resize(shape_function_container.NumberOfNonzeroControlPoints(), 3);
+
+        for (IndexType i = 0; i < shape_function_container.NumberOfNonzeroControlPoints(); ++i) {
+            rResult(i, 0) = shape_function_container(i, 1);
+            rResult(i, 1) = shape_function_container(i, 2);
+            rResult(i, 2) = shape_function_container(i, 3);
```

Comment:
Here is some more consistency with the nurbs_surface. The indices in rResults are (0, i) ... , but I am pretty sure they should be (i,0)...

Reply:
unfortunately the way that rResult and the container is defined is different. We cannot touch the definition of the rResult. So you should change it in the way that all other geometries are defined.

Reply:
https://github.com/KratosMultiphysics/Kratos/blob/0e8413b222cb1c091b3e2f45da627fd26efa7794/kratos/geometries/geometry.h#L2600
In the base_geometry the second argument "m" is the local space dimension. So rResults(i,0); (i,1); (i,2) should be correct.
However in the surface it is:
https://github.com/KratosMultiphysics/Kratos/blob/0e8413b222cb1c091b3e2f45da627fd26efa7794/kratos/geometries/nurbs_surface_geometry.h#L696

Reply:
This would be unfortunate...